RockWeb/App_Code/FileUploader.ashx.cs

```diff
@@ -133,54 +133,82 @@ private void ProcessContentFile( HttpContext context, HttpPostedFile uploadedFil
-            var filename = Path.GetFileName( uploadedFile.FileName );
-            filename = Regex.Replace( filename, @"[<>*%&:\\]", string.Empty, RegexOptions.CultureInvariant );
 
-            // get folderPath and construct filePath
-            string relativeFolderPath = context.Request.Form["folderPath"] ?? string.Empty;
-            string relativeFilePath = Path.Combine( relativeFolderPath, filename );
-            string rootFolderParam = context.Request.QueryString["rootFolder"];
+            // NEVER TRUST THE CLIENT!!!!
+            string untrustedFileName = uploadedFile.FileName;
+            string untrustedFolderPath = context.Request.Form["folderPath"] ?? string.Empty;
+            string encryptedRootFolder = context.Request.QueryString["rootFolder"];
 
-            string rootFolder = string.Empty;
+        
+            /* Scrub the file name */
 
-            if ( !string.IsNullOrWhiteSpace( rootFolderParam ) )
+            string scrubedFileName = ScrubFileName( untrustedFileName );
+
+
+            /* Determine the root upload folder */
+
+            string trustedRootFolder = string.Empty;
+
+            // If a rootFolder was specified in the URL, try decrypting it (It is encrypted to help prevent direct access to filesystem).
+            if ( !string.IsNullOrWhiteSpace( encryptedRootFolder ) )
             {
-                // if a rootFolder was specified in the URL, decrypt it (it is encrypted to help prevent direct access to filesystem)
-                rootFolder = Rock.Security.Encryption.DecryptString( rootFolderParam );
+                trustedRootFolder = Encryption.DecryptString( encryptedRootFolder );
             }
 
-            if ( string.IsNullOrWhiteSpace( rootFolder ) )
+            // If we don't have a rootFolder, default to the ~/Content folder.
+            if ( string.IsNullOrWhiteSpace( trustedRootFolder ) )
             {
-                // set to default rootFolder if not specified in the params
-                rootFolder = "~/Content";
+                trustedRootFolder = "~/Content";
             }
 
-            string physicalRootFolder = context.Request.MapPath( rootFolder );
-            string physicalContentFolderName = Path.Combine( physicalRootFolder, relativeFolderPath.TrimStart( new char[] { '/', '\\' } ) );
 
-            // Make sure the physicalContentFolderName doesn't have any special directory navigation indicators
-            if ( physicalContentFolderName != System.IO.Path.GetFullPath( physicalContentFolderName ) )
+            /* Combine the root and folder paths to get the real physical location */
+
+            // Get the absolute path for our trusted root.
+            string trustedPhysicalRootFolder = Path.GetFullPath( context.Request.MapPath( trustedRootFolder ) );
+
+            // Treat rooted folder paths as relative
+            string untrustedRelativeFolderPath = untrustedFolderPath.TrimStart( Path.GetPathRoot( untrustedFolderPath ).ToCharArray() );
```

Comment:
I get an exception here when untrustedFolderPath = "".  To reproduce this issue, use the FileUploader (ContentFileMode) in the Rock ControlGallery.